File: Java/Data Structures and Algorithms/Recursion/MergeSort.java

class DArray {
    private long[] theArray;
    private int nItems;

    public DArray (int max) {
        theArray = new long[max];
        nItems = 0;
    }

    public void insert (long value) {
        theArray[nItems++] = value;
    }

    public void display () {
        for (int i = 0 ; i < nItems ; ++i)
            System.out.print (theArray[i] + " ");
        System.out.println ("");
    }

    public void mergeSort () {
        long[] workspace = new long[nItems];
        recMergeSort (workspace, 0, nItems-1);
    }

    private void recMergeSort (long[] workSpace, int lowerBound, int upperBound) {
        if (lowerBound == upperBound) // if range is 1
            return;
        else {
            int mid = (lowerBound+upperBound)/2;
            recMergeSort (workSpace, lowerBound, mid); // sort low half
            recMergeSort (workSpace, mid+1, upperBound); // sort high half
            merge (workSpace, lowerBound, mid+1, upperBound); // merge them
        }
    }

    private void merge(long[] workSpace, int lowPtr, int highPtr, int upperBound) {
        int i = 0; // workspace index
        int lowerBound = lowPtr;
        int mid = highPtr-1;
        int n = upperBound-lowerBound+1; // number of items

        while (lowPtr <= mid && highPtr <= upperBound) {
            if (theArray[lowPtr] < theArray[highPtr])
                workSpace[i++] = theArray[lowPtr++];
            else
                workSpace[i++] = theArray[highPtr++];
        }

        while (lowPtr <= mid)
            workSpace[i++] = theArray[lowPtr++];

        while (highPtr <= upperBound)
            workSpace[i++] = theArray[highPtr++];

        for (i = 0 ; i < n ; ++i)
            theArray[lowerBound+i] = workSpace[i];
    }
}


public class MergeSort {
    public static void main (String[] args) {
        int maxSize = 100;
        DArray arr = new DArray (maxSize);

        arr.insert (64);
        arr.insert (21);
        arr.insert (33);
        arr.insert (70);
        arr.insert (12);
        arr.insert (85);
        arr.insert (44);
        arr.insert (3);
        arr.insert (99);
        arr.insert (0);
        arr.insert (108);
        arr.insert (36);

        arr.display ();

        arr.mergeSort ();

        arr.display ();
    }
}

File: C/algorithms/sorting/mergesort.c
#include <stdio.h>
#include <stdlib.h>

void merge (int a[], int low, int mid, int high) {
  int b[10000];
  int i = low, j = mid + 1, k = 0;

  while (i <= mid && j <= high) {
    if (a[i] <= a[j])
      b[k++]  = a[i++];
    else
      b[k++] = a[j++];
  }
  while (i <= mid)
    b[k++] = a[i++];

  while (j <= high)
    b[k++] = a[j++];

  k--;
  while (k >= 0) {
    a[low + k] = b[k];
    k--;
  }
}

void mergeSort (int a[], int low, int high) {
  if (low < high) {
    int m = (high+low)/2;
    mergeSort(a,low,m);
    mergeSort(a,m+1,high);
    merge(a,low,m,high);
  }
}

void printArray(int A[], int size) {
    int i;
    for (i=0; i < size; i++)
        printf("%d ", A[i]);
    printf("\n");
}

int main() {
  int arr[] = {12, 11, 13, 5, 6, 7, 9};
    int arr_size = sizeof(arr)/sizeof(arr[0]);

    printf("Given array is \n");
    printArray(arr, arr_size);

    mergeSort(arr, 0, arr_size - 1);

    printf("\nSorted array is \n");
    printArray(arr, arr_size);
    return 0;
}
